File: packages/core/src/platform/services/context.ts
import type { DomainEvent, WorkflowStats, WorkspaceStats, WorkspaceId } from "@agently/contracts";
import { ids } from "@agently/contracts";
import type { Repositories } from "../repositories/types.js";
import type { EventBus } from "../events/bus.js";
import type { Clock } from "../clock.js";
import type { Logger } from "../logger.js";

/** Everything the services need, injected once at platform construction. */
export interface PlatformDeps {
  repos: Repositories;
  bus: EventBus;
  clock: Clock;
  logger: Logger;
  /** Materialized rollups from the seed (a stand-in for a future stats table). */
  seedStats: {
    workflowStats: Record<string, WorkflowStats>;
    workspaceStats: WorkspaceStats;
  };
}

/** Distributive Omit so each union member keeps its own payload fields. */
type DistributiveOmit<T, K extends keyof never> = T extends unknown ? Omit<T, K> : never;
export type EmitInput = DistributiveOmit<DomainEvent, "id" | "occurredAt" | "workspaceId">;

/** Build a publisher that stamps every event with id/time/tenant. */
export function createEmitter(bus: EventBus, clock: Clock, workspaceId: WorkspaceId) {
  return (event: EmitInput): void => {
    bus.publish({
      id: ids.event(),
      occurredAt: clock.iso(),
      workspaceId,
      ...event,
    } as DomainEvent);
  };
}

export type Emitter = ReturnType<typeof createEmitter>;
// re-exported for convenience at call sites
export { ids };

File: apps/api/internal/services/context.go
// Package services holds the durable business logic (use-cases). Each service
// depends only on repository interfaces, the event bus, the clock, and the
// logger — never on a concrete store — mirroring
// packages/core/src/platform/services.
package services

import (
	"github.com/agently/api/internal/domain"
	"github.com/agently/api/internal/platform"
)

// SeedStats are the materialized rollups from the seed (a stand-in for a future
// stats table).
type SeedStats struct {
	WorkflowStats  map[string]domain.WorkflowStats
	WorkspaceStats domain.WorkspaceStats
}

// Deps is everything the services need, injected once at platform construction.
type Deps struct {
	Repos     *platform.Repositories
	Bus       platform.EventBus
	Clock     platform.Clock
	Logger    platform.Logger
	SeedStats SeedStats
}

// Emitter publishes a domain event after stamping its envelope (id, occurredAt,
// workspaceId) and its discriminator type. Mirrors createEmitter in context.ts.
type Emitter func(event domain.DomainEvent)

// NewEmitter builds an emitter bound to the bus, clock, and workspace.
func NewEmitter(bus platform.EventBus, clock platform.Clock, workspaceID domain.WorkspaceId) Emitter {
	return func(event domain.DomainEvent) {
		stamped := stampEvent(event, domain.NewEventId(), domain.Timestamp(clock.ISO()), workspaceID)
		bus.Publish(stamped)
	}
}

// stampEvent fills the envelope + type discriminator on a partially-built event.
// Each event is built by the services with only its payload fields set; this
// fills id/occurredAt/workspaceId/type so the wire shape matches the contract.
func stampEvent(e domain.DomainEvent, id domain.EventId, at domain.Timestamp, ws domain.WorkspaceId) domain.DomainEvent {
	base := domain.EventBase{ID: id, OccurredAt_: at, WorkspaceID_: ws}
	switch ev := e.(type) {
	case domain.RunQueuedEvent:
		ev.EventBase, ev.Type = base, ev.EventType()
		return ev
	case domain.RunStartedEvent:
		ev.EventBase, ev.Type = base, ev.EventType()
		return ev
	case domain.RunProgressEvent:
		ev.EventBase, ev.Type = base, ev.EventType()
		return ev
	case domain.RunAgentTransitionedEvent:
		ev.EventBase, ev.Type = base, ev.EventType()
		return ev
	case domain.RunLogAppendedEvent:
		ev.EventBase, ev.Type = base, ev.EventType()
		return ev
	case domain.RunBrowserActionEvent:
		ev.EventBase, ev.Type = base, ev.EventType()
		return ev
	case domain.RunArtifactProducedEvent:
		ev.EventBase, ev.Type = base, ev.EventType()
		return ev
	case domain.RunCostThresholdEvent:
		ev.EventBase, ev.Type = base, ev.EventType()
		return ev
	case domain.RunFinishedEvent:
		ev.EventBase, ev.Type = base, ev.EventType()
		return ev
	case domain.NotificationCreatedEvent:
		ev.EventBase, ev.Type = base, ev.EventType()
		return ev
	default:
		return e
	}
}
